packages/grafana-ui/src/components/GraphNG/GraphNG.tsx

```diff
@@ -55,310 +33,182 @@ export interface GraphNGProps {
-
-      // The builder will manage unique scaleKeys and combine where appropriate
-      builder.addScale({
-        scaleKey,
-        orientation: ScaleOrientation.Vertical,
-        direction: ScaleDirection.Up,
-        distribution: customConfig.scaleDistribution?.type,
-        log: customConfig.scaleDistribution?.log,
-        min: field.config.min,
-        max: field.config.max,
-        softMin: customConfig.axisSoftMin,
-        softMax: customConfig.axisSoftMax,
-      });
-
-      if (customConfig.axisPlacement !== AxisPlacement.Hidden) {
-        builder.addAxis({
-          scaleKey,
-          label: customConfig.axisLabel,
-          size: customConfig.axisWidth,
-          placement: customConfig.axisPlacement ?? AxisPlacement.Auto,
-          formatValue: (v) => formattedValueToString(fmt(v)),
-          theme,
-        });
+    if (data !== prevProps.data) {
+      if (!alignedDataFrame) {
+        return;
       }
 
-      const showPoints = customConfig.drawStyle === DrawStyle.Points ? PointVisibility.Always : customConfig.showPoints;
-
-      let { fillOpacity } = customConfig;
-      if (customConfig.fillBelowTo) {
-        if (!indexByName) {
-          indexByName = getNamesToFieldIndex(frame);
-        }
-        const t = indexByName.get(getFieldDisplayName(field, frame));
-        const b = indexByName.get(customConfig.fillBelowTo);
-        if (isNumber(b) && isNumber(t)) {
-          builder.addBand({
-            series: [t, b],
-            fill: null as any, // using null will have the band use fill options from `t`
-          });
-        }
-        if (!fillOpacity) {
-          fillOpacity = 35; // default from flot
-        }
-      }
+      const hasStructureChanged = !compareArrayValues(data, prevProps.data, compareDataFrameStructures);
 
-      builder.addSeries({
-        scaleKey,
-        showPoints,
-        colorMode,
-        fillOpacity,
-        theme,
-        drawStyle: customConfig.drawStyle!,
-        lineColor: customConfig.lineColor ?? seriesColor,
-        lineWidth: customConfig.lineWidth,
-        lineInterpolation: customConfig.lineInterpolation,
-        lineStyle: customConfig.lineStyle,
-        barAlignment: customConfig.barAlignment,
-        pointSize: customConfig.pointSize,
-        pointColor: customConfig.pointColor ?? seriesColor,
-        spanNulls: customConfig.spanNulls || false,
-        show: !customConfig.hideFrom?.graph,
-        gradientMode: customConfig.gradientMode,
-        thresholds: config.thresholds,
-
-        // The following properties are not used in the uPlot config, but are utilized as transport for legend config
-        dataFrameFieldIndex: field.state?.origin,
-        fieldName: getFieldDisplayName(field, frame),
-        hideInLegend: customConfig.hideFrom?.legend,
-      });
+      if (shouldConfigUpdate || hasStructureChanged) {
+        const builder = preparePlotConfigBuilder(alignedDataFrame, theme, this.getTimeRange, this.getTimeZone);
```

Comment:
this function still seems like the weak link -- any functions used in config need wrapped functions around them -- rather than just using this.props